# c++中的异常机制之异常的类型和声明周期

* 1.throw的异常是有类型的，可以是数字、字符串、类对象。
* 2.throw的异常是有类型的，catch严格按照类型匹配。
* 3.注意异常的内存模型。
* 4.对于基础的数据类型，catch完毕之后，异常的声明周期就会结束。

**如果是类和对象的时候，建议使用引用**


```
#include <iostream>

using namespace std;

class BadProcessType {
public:
    BadProcessType() {
        cout << "BadProcessType构造函数do \n";
    }

    BadProcessType(const BadProcessType &obj) {
        cout << "BadProcessType copy构造函数do \n";
    };

    ~BadProcessType() {
        cout << "BadProcessType析构函数do \n";
    }
};

void my_strcpy3(char *to, char *from) {

    if (*from == 'a') {
        printf("开始 BadProcessType类型异常 \n");
        throw BadProcessType();
    }

    if (*from == 'c') {
        throw new BadProcessType;
    }
    while (*from != '\0') {
        *to = *from;
        to++;
        from++;
    }
    *to = '\0';
}

int main() {
    int ret = 0;
    char buf1[] = "c";
    char buf2[1024] = {0};

    try {
        my_strcpy3(buf2, buf1);
    }
//    catch(BadProcessType e){  //调用拷贝构造函数
//        cout << "类型异常" << endl;
//    }
    catch(BadProcessType *e){
        cout << "类型异常" << endl;
        delete e;  //指针需要手动的去delete.
    }
//    catch(BadProcessType &e){  //引用指向，不会调用拷贝构造
//        cout << "类型异常" << endl;
//    }
    return 0;
}
```
